core/client/fs/src/main/java/alluxio/client/block/stream/LocalFilePacketWriter.java

```diff
@@ -201,10 +225,91 @@ private void ensureReserved(long pos) throws IOException {
       return;
     }
     long toReserve = Math.max(pos - mPosReserved, FILE_BUFFER_BYTES);
-    NettyRPC.call(mNettyRPCContext, new ProtoMessage(
-        mCreateRequest.asLocalBlockCreateRequest().toBuilder().setSpaceToReserve(toReserve)
-            .setOnlyReserveSpace(true).build()));
+    mRequestObserver.onNext(mCreateRequest.toBuilder().setSpaceToReserve(toReserve)
+          .setOnlyReserveSpace(true).build());
+    try (LockResource lr = new LockResource(mResponseObserver.getLock())) {
+      if (!mResponseObserver.getReservedOrFailed()
+          .await(READ_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
+        throw new DeadlineExceededException(String.format(
+            "Timeout reserving space for request %s.", mCreateRequest.toString()));
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(String.format(
+          "Interrupted reserving space for request %s.", mCreateRequest.toString()), e);
+    }
     mPosReserved += toReserve;
   }
+
+  private static class ResponseObserver implements StreamObserver<CreateLocalBlockResponse> {
+    private final ReentrantLock mLock = new ReentrantLock();
+    /** This condition is met if mError != null or create response is returned. */
+    private final Condition mCreatedOrFailed = mLock.newCondition();
```

Comment:
Do we need all of these condition variables just to capture the state of the rpc?

Reply:
Yes, due to the async nature of the gRPC bi-directional stream call we need to wait on the variables to convert it to a blocking call.